### PR TITLE
[New Hack] 054-Azure Load Testing

### PR DESCRIPTION
**Pull request by Whowong**
New Hack Submission - Azure Load Testing

@kevinmgates

### CODE REVIEW

**Comment by kevinmgates** [comment]
Nicely done, @Whowong

File: 054-AzureLoadTesting/Student/Challenge-07.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Challenge 07 - Load Testing With Chaos Experiment (Resilience Testing)
+
+[< Previous Challenge](./Challenge-06.md) - **[Home](../README.md)**
+
+## Introduction
+
+We have gone over load testing for the majority of this hack.  While the information you gather from load testing can help determine how you scale your application.  The resiliency of an application isn’t solely decided by how quickly it can scale up or out - but also how it handles failures. This means planning for a failure of every application component: a single container, cluster, VM, database and region.  Testing while having certain components fail is called Resilience testing, and we will be going over that here.
```
Comment (perktime) [inline 054-AzureLoadTesting/Student/Challenge-07.md:7]: Change to "Testing while having certain components fail is called Resilience testing. We will be going over that here."

**Review by perktime — CHANGES_REQUESTED** [review]

File: 054-AzureLoadTesting/Coach/Solution-01.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+- Identify the load characteristics and scenario
+    - Our current average daily load is 1000 users/hour.  We will mimic the average daily load hitting all 3 endpoints.  Since there is consistent daily load we will mimic this with gradual load increasing linearly.  
+
+- Identify the test failure criteria
+    - Our test will fail if we are unable to support 1000 users as we already know this is the current demand.
+    - We will also consider this a failed test if we find performance below expected thresholds as this may impact customer satisfaction.
+
+- Identify how you will be monitoring your application
+    - We will be monitoring our applications with application insights to detect any errors and monitor performance.
```
Comment (jordanbean-msft) [inline 054-AzureLoadTesting/Coach/Solution-01.md:29]: capitalize Application Insights

File: 054-AzureLoadTesting/Coach/Solution-07.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# Challenge 07 - Load Testing With Chaos Experiment (Resilience Testing) - Coach's Guide 
+
+[< Previous Solution](./Solution-06.md) - **[Home](./README.md)**
+
+## Notes & Guidance
+
+Here are instructions on how to setup Chaos Studio and run the experiment.  Chaos studio does not offer a GitHub action at this time.  We want the users to get a taste of chaos engineering so they can trigger the chaos experiment manually while the load test runs.  They are not required to implement this all in the pipeline.  That is an advanced challenge.
```
Comment (jordanbean-msft) [inline 054-AzureLoadTesting/Coach/Solution-07.md:7]: capitalize Chaos Studio & GitHub Actions, do this throughout

File: 054-AzureLoadTesting/Coach/Solution-07.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        - Go to your subscription
+        - On the left-hand side, select "Resource provider"
+        - In the list of providers, search for "Microsoft.Chaos"
+        - Click on the provider and select "Register"
+    - Go to Azure Chaos Studio
+    - Under "Targets" select "Enable Targets" then "Enable service-direct targets"
+    - Select the Cosmos DB service for your load test.
+    - On the left-hand side, select "Experiments"
+    - Select "Add an experiment"
+    - Fill in your subscription, resource group, location, and name for this experiment.  Keep track of the experiment name as a managed user will be created for you.
```
Comment (jordanbean-msft) [inline 054-AzureLoadTesting/Coach/Solution-07.md:22]: capitalize Subscription & Resource Group, do this throughout

File: 054-AzureLoadTesting/Student/Challenge-00.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+## Description
+
+Now that you have the common prerequisites installed on your workstation, there is a sample application that we will be working off of for this hack.
+
+Your coach will provide you with a Resources.zip file that contains resources you will need to complete the hack. If you plan to work locally, you should unpack it on your workstation. If you plan to use the Azure Cloud Shell, you should upload it to the Cloud Shell and unpack it there.
+
+### Sample WebApp with Cosmos DB 
+ 
+ The sample app is a WebApp deployed on app service with cosmos db as database. It counts the number of visitors visiting the page and inserts the same into a sample collection in Cosmos DB.
```
Comment (jordanbean-msft) [inline 054-AzureLoadTesting/Student/Challenge-00.md:32]: Capitalize App Service

**Review by jordanbean-msft — CHANGES_REQUESTED** [review]

File: 054-AzureLoadTesting/Coach/README.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+- Challenge 01: **[Develop a Load Testing Strategy](./Solution-01.md)**
+	 - How to develop a load testing strategy for your application
+- Challenge 02: **[Create a Load Testing Script](./Solution-02.md)**
+	 - Deploy a sample application and create JMeter scripts to support your load testing strategy
+- Challenge 03: **[Create Azure Load Testing Service and Establish Baselines](./Solution-03.md)**
+	 - Create Azure Load Testing Service and learn techniques on how to establish baselines for your application
+- Challenge 04: **[Enable Automated Load Testing (CI/CD)](./Solution-04.md)**
+	 - Incorporating load testing into your CI/CD Pipeline
+- Challenge 05: **[Identify & Remediate Bottlenecks](./Solution-05.md)**
+	 - Reviewing load test results and identifying bottlenecks
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/README.md]: Done

File: 054-AzureLoadTesting/Coach/README.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+- Challenge 06: **[Stress Testing](./Solution-06.md)**
+	 - How to perform stress tests and observing your application behavior
+- Challenge 07: **[Load Testing With Chaos Experiment (Resilience Testing)](./Solution-07.md)**
+	 - Incorporating load testing and chaos experiments together
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/README.md]: Done

File: 054-AzureLoadTesting/Coach/Solution-01.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+Below is a sample load testing plan we have mapped out.  Students will have something different as we left it up to the student to interpret the environment however the overall feel should be similar.
+
+- Sample Load Testing Plan
+
+
+- Define what services and the scope of your testing - Requirements
+    - We will be simulating an average day when a user hits our website during normal business hours 8 - 5 PM.  We will be testing all 3 endpoints (Get/Add/lasttimestamp) to ensure all features are tested.  
+
+- Environment Requirements
+    - Since we are testing the user we will need the following resources at the same level as production.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-01.md]: Done

File: 054-AzureLoadTesting/Coach/Solution-01.md (hunk 2 above)
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-01.md]: Done

File: 054-AzureLoadTesting/Coach/README.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+- Challenge 00: **[Prerequisites - Ready, Set, GO!](./Solution-00.md)**
+	 - Prepare your workstation to work with Azure.
+- Challenge 01: **[Develop a Load Testing Strategy](./Solution-01.md)**
+	 - How to develop a load testing strategy for your application
+- Challenge 02: **[Create a Load Testing Script](./Solution-02.md)**
+	 - Deploy a sample application and create JMeter scripts to support your load testing strategy
+- Challenge 03: **[Create Azure Load Testing Service and Establish Baselines](./Solution-03.md)**
+	 - Create Azure Load Testing Service and learn techniques on how to establish baselines for your application
+- Challenge 04: **[Enable Automated Load Testing (CI/CD)](./Solution-04.md)**
+	 - Incorporating load testing into your CI/CD Pipeline
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/README.md]: Done

File: 054-AzureLoadTesting/Coach/Solution-01.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    
+- Identify potential bottlenecks/limitations in advance
+    - Since we do not cache any data at any level.  Any component failing could cause an issue downstream.
+    
+- Please note any assumptions
+    - We are assuming these are only casual end users and not other 3rd parties who could be trying to obtain information about our traffic.
+    - We are using consumption based model for Azure App Services
+    - Application is not caching any data and always makes calls to Cosmos DB
+
+- Reinforce the importance of using existing metrics/usage patterns/data to inform the load testing own their own application.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-01.md]: Done

File: 054-AzureLoadTesting/Coach/Solution-03.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+## Notes & Guidance
+
+Below are the instructions to complete challenge 3
+
+- Create Azure Load Testing resource
+    - Search for the Azure Load Testing resource on the top search bar and select the resource.
+    - On the top left hand corner select "Create"
+    - Select the Resource Group that was created by the sample app or create your own.  The resources must be in the same location.
+    - Enter a unique name for the resource.
+    - Ensure the location matches the location of your application deployed to.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-03.md]: Done

File: 054-AzureLoadTesting/Coach/Solution-03.md [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    - Select "Review + create"
+    - Review your configurations and then select "Create"
+- Create Load Test
+    - Go into your new resource.  You may see a warning on the top depending on your current access describing the access levels required.  If so add yourself as one of the roles such as "Load Testing Owner" or "Load Testing Contributor" as you will not be able to proceed otherwise.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-03.md]: Done

File: 054-AzureLoadTesting/Coach/Solution-03.md [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    - Select "Create"
+    - Fill out the name of the test and a description of the test.
+    - Go to the "Test Plan" tab and select the JMeter script you have locally and then select "Upload"
+    - Under the "Parameters" tab, enter "webapp" for the name of the environment variable.  Followed by the URL for your sample app in the "Value" section.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-03.md]: Done

File: 054-AzureLoadTesting/Coach/Solution-04.md [hunk 14]
```diff
@@ -0,0 +1,14 @@
+# Challenge 04 - Enable Automated Load Testing (CI/CD) - Coach's Guide 
+
+[< Previous Solution](./Solution-03.md) - **[Home](./README.md)** - [Next Solution >](./Solution-05.md)
+
+## Notes & Guidance
+
+
+High level Notes
+
+- There is a sample action which will run a load test.  The default action will either create or run the existing load test.  It is case-sensitive and the GitHub action looks like it lowercase all the tests.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-04.md]: Thanks, clarified it as it was confusing.

File: 054-AzureLoadTesting/Coach/Solution-05.md [hunk 15]
```diff
@@ -0,0 +1,13 @@
+# Challenge 05 - Identify & Remediate Bottlenecks - Coach's Guide 
+
+[< Previous Solution](./Solution-04.md) - **[Home](./README.md)** - [Next Solution >](./Solution-06.md)
+
+## Notes & Guidance
+- This link goes into the issues in detail
+https://docs.microsoft.com/en-us/azure/load-testing/tutorial-identify-bottlenecks-azure-portal
+
+- The students should be seeing many errors and that cosmos db RU are maxed out.  If they are not, have them increase the amount of load in their tests.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-05.md]: Done

File: 054-AzureLoadTesting/Coach/Solution-05.md [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+- Make sure the students have configured monitoring integration from azure load test with the services such as App Service, App Insights, and Cosmos DB.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-05.md]: Done

File: 054-AzureLoadTesting/Coach/Solution-07.md [hunk 17]
```diff
@@ -0,0 +1,37 @@
+# Challenge 07 - Load Testing With Chaos Experiment (Resilience Testing) - Coach's Guide 
+
+[< Previous Solution](./Solution-06.md) - **[Home](./README.md)**
+
+## Notes & Guidance
+
+Here are instructions on how to setup Chaos Studio and run the experiment.  Chaos studio does not offer a GitHub action at this time.  We want the users to get a taste of chaos engineering so they can trigger the chaos experiment manually while the load test runs.  They are not required to implement this all in the pipeline.  That is an advanced challenge.
+
+- Azure Chaos studio only supports a number of faults such as VMs, AKS, Cosmos DB, and networking to name a few.  Since our sample app only contains App Services and Cosmos DB, we will use the Cosmos DB faults.
+- First you will need to enable geo-redundancy in your Cosmos DB.  Go to your Cosmos DB instance and select "Enable Geo Redundancy" on the top bar.  Keep note of the paired region your cosmos db is paired with.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-07.md]: Done

File: 054-AzureLoadTesting/README.md [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+	 - Incorporating load testing into your CI/CD Pipeline
+- Challenge 05: **[Identify & Remediate Bottlenecks](Student/Challenge-05.md)**
+	 - Reviewing load test results and identifying bottlenecks
+- Challenge 06: **[Stress Testing](Student/Challenge-06.md)**
+	 - How to perform stress tests and observing your application behavior
+- Challenge 07: **[Load Testing With Chaos Experiment (Resilience Testing)](Student/Challenge-07.md)**
+	 - Incorporating load testing and chaos experiments together
+
+## Prerequisites
+- Familiarity with the fundamentals of load testing.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/README.md]: Done

File: 054-AzureLoadTesting/Student/Challenge-00.md [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+- [Visual Studio Code](../../000-HowToHack/WTH-Common-Prerequisites.md#visual-studio-code)
+- [Azure Storage Explorer](../../000-HowToHack/WTH-Common-Prerequisites.md#azure-storage-explorer)
+- [Apache JMeter](https://jmeter.apache.org/usermanual/get-started.html) - This requires Java 8+ to be installed
+- [GitHub](https://github.com/) or [Azure DevOps](https://azure.microsoft.com/en-us/services/devops/?nav=min)
+
+## Description
+
+Now that you have the common prerequisites installed on your workstation, there is a sample application that we will be working off of for this hack.
+
+Your coach will provide you with a Resources.zip file that contains resources you will need to complete the hack. If you plan to work locally, you should unpack it on your workstation. If you plan to use the Azure Cloud Shell, you should upload it to the Cloud Shell and unpack it there.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Student/Challenge-00.md]: Done

File: 054-AzureLoadTesting/Student/Challenge-00.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Student/Challenge-00.md]: Done

File: 054-AzureLoadTesting/Student/Challenge-00.md [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+        az group delete --name myResourceGroup
+
+Similarly, you can utilize the **Delete resource group** toolbar button on the sample application's resource group to remove all the resources.
+
+## Success Criteria
+
+To complete this challenge successfully, you should be able to:
+
+- Verify that you have a bash shell with the Azure CLI available.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Student/Challenge-00.md]: Done

File: 054-AzureLoadTesting/Student/Challenge-02.md [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+## Learning Resources
+
+- [Node.js sample application](https://github.com/Whowong/nodejs-appsvc-loadtest)
+- [Apache JMeter Docs](https://jmeter.apache.org/index.html)
+- [Create Test Plan from Template](https://jmeter.apache.org/usermanual/get-started.html#template)
+- [JMeter best practices](https://jmeter.apache.org/usermanual/best-practices.html)
+- [Custom Plugins for Apache JMeter](https://jmeter-plugins.org/)
+
+## Advanced Challenges (Optional)
+- Your initial load test scripts may gradually increase load in a linear rate.  What if you wanted to simulate users starting their day at work at different time zones.  See if you are able to create a load test which steps into higher load like a stairway.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Student/Challenge-02.md]: Done

File: 054-AzureLoadTesting/Student/Challenge-03.md [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+[< Previous Challenge](./Challenge-02.md) - **[Home](../README.md)** - [Next Challenge >](./Challenge-04.md)
+
+## Pre-requisites
+
+- Before creating your Azure Load Test service, ensure you have created your load testing script(s) in the previous challenge.
+## Introduction
+
+In this challenge you will create the Load Testing service in Azure and use the load testing scripts you created in our last challenge to run baseline tests.
+
+Baselines help to determine the current efficiency state of your application and its supporting infrastructure. They can provide good insights for improvements and determine if the application is meeting business goals. Lastly, they can be created for any application regardless of its maturity. No matter when you establish the baseline, measure performance against that baseline during continued development. When code and, or infrastructure changes, the effect on performance can be actively measured.
```
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Student/Challenge-03.md]: Done

File: 054-AzureLoadTesting/Student/Challenge-07.md (hunk 1 above)
Comment (Whowong) [reply 054-AzureLoadTesting/Student/Challenge-07.md]: Done

File: 054-AzureLoadTesting/Coach/Solution-07.md (hunk 3 above)
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/Solution-07.md]: Done

File: 054-AzureLoadTesting/Student/Challenge-00.md (hunk 5 above)
Comment (Whowong) [reply 054-AzureLoadTesting/Student/Challenge-00.md]: Done

File: 054-AzureLoadTesting/README.md [hunk 23]
```diff
@@ -0,0 +1,42 @@
+# What The Hack - xxx-AzureLoadTesting
```
Comment (jrzyshr) [inline 054-AzureLoadTesting/README.md:1]: Remove the "xxx-" and add spaces between Azure Load Testing.

File: 054-AzureLoadTesting/Coach/README.md [hunk 24]
```diff
@@ -0,0 +1,70 @@
+# What The Hack - Azure Load Testing - Coach Guide
+
+## Introduction
+Welcome to the coach's guide for the Azure Load Testing What The Hack. Here you will find links to specific guidance for coaches for each of the challenges.  
+
+NOTE: If you are a Hackathon participant, this is the answer guide. Don't cheat yourself by looking at these during the hack! Go learn something. :)
+## Duration
```
Comment (jrzyshr) [inline 054-AzureLoadTesting/Coach/README.md:7]: This belongs under the new template's "Proposed Agenda" section!

File: 054-AzureLoadTesting/Coach/README.md [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+This hack is estimated to take 3 days with 4 hours each day.
+
+## Preparation
```
Comment (jrzyshr) [inline 054-AzureLoadTesting/Coach/README.md:10]: This belongs under the new "Coach Prerequisites" section in the template!  I don't think it needs its own section header.  Just add the content where the template says you should.

**Review by jrzyshr — CHANGES_REQUESTED** [review]
I haven't made a full pass yet, but wanted to get these few minor comments in.  I'll trust Jordan and Pete for the rest.

File: 054-AzureLoadTesting/Coach/README.md (hunk 24 above)
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/README.md]: Done

File: 054-AzureLoadTesting/Coach/README.md (hunk 25 above)
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/README.md]: Done

File: 054-AzureLoadTesting/Coach/README.md (hunk 24 above)
Comment (jrzyshr) [reply 054-AzureLoadTesting/Coach/README.md]: Now that I've re-read this, I think the boilerplate text about needing to package /Student/Resources covers this?  I would either remove your sentence all together, or maybe modify/add a line to the "Student Resources" section in the template.

Comment (Whowong) [reply 054-AzureLoadTesting/Coach/README.md]: Yeah that makes sense.  I just removed it.

File: 054-AzureLoadTesting/Coach/README.md [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+	- Azure Load Testing
+	- Azure App Service
+	- Azure Cosmos DB
+	- Azure Application Insights
+
+## Suggested Hack Agenda
+This hack is estimated to take 3 days with 4 hours each day.
+## Repository Contents
+
+_The default files & folders are listed below. You may add to this if you want to specify what is in additional sub-folders you may add._
```
Comment (jrzyshr) [inline 054-AzureLoadTesting/Coach/README.md:57]: This line in italics should be removed.  It is template instructions for the author. 

Given my previous comment about removing the redundant line about packaging up the sample app, you might want to add something in this "Repository Contents" section explaining what IS in the /Student/Resource folder??

**Review by jrzyshr — CHANGES_REQUESTED** [review]

File: 054-AzureLoadTesting/Coach/README.md (hunk 26 above)
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/README.md]: Done

**Review by jordanbean-msft — APPROVED** [review]

File: 054-AzureLoadTesting/Coach/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (perktime) [reply 054-AzureLoadTesting/Coach/README.md]: Oops. I missed telling you to change identifying to "identify"

File: 054-AzureLoadTesting/Student/Challenge-00.md [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+ The sample app is a WebApp deployed on app service with cosmos db as database. It counts the number of visitors visiting the page and inserts the same into a sample collection in Cosmos DB.
+
+### Installation
+
+1. In your terminal window, log into Azure and set a subscription(subscription which would contain the webapp) :
+
+        az login
+        az account set -s mySubscriptionName
+
+1. Deploy the sample app using the PowerShell script. (Tip: macOS users can install PowerShell [here](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-macos?view=powershell-7.1))
```
[the remark this answers is not in the document]
Comment (perktime) [reply 054-AzureLoadTesting/Student/Challenge-00.md:41]: We will work on converting this but I will resolve this for now

File: 054-AzureLoadTesting/Coach/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Whowong) [reply 054-AzureLoadTesting/Coach/README.md]: Done

**Review by perktime — APPROVED** [review]

**Review by jrzyshr — APPROVED** [review]

**Comment by jrzyshr** [comment]
This PR closes Issue #348

**Pull request merged**